quinn-proto/src/connection/mod.rs

```diff
@@ -54,10 +54,43 @@ use timer::{Timer, TimerTable};
 
 /// Protocol state and logic for a single QUIC connection
 ///
-/// Objects of this type receive `ConnectionEvent`s and emit `EndpointEvents` and application
-/// `Event`s to make progress. To handle timeouts, a `Connection` returns timer updates and
+/// Objects of this type receive [`ConnectionEvent`]s and emit [`EndpointEvents`] and application
+/// [`Event`]s to make progress. To handle timeouts, a `Connection` returns timer updates and
 /// expects timeouts through various methods. A number of simple getter methods are exposed
 /// to allow callers to inspect some of the connection state.
+///
+/// `Connection` has roughly 4 types of methods:
+///
+/// - A. Simple getters, taking `&self`
+/// - B. Handlers for incoming events from the network or system, named `handle_*`.
+/// - C. State machine mutators, for incoming commands from the application. For
+///   convenience we refer to this as "performing I/O" below, however as per
+///   the design of this library none of the functions actually perform
+///   system-level I/O. For example, [`read`](self::read) and [`write`](self::write),
+///   but also things like [`reset`](self::reset).
+/// - D. Polling functions for outgoing events or results, named `poll_*`.
+///
+/// The simplest way to use this API correctly is to call (B) and (C) whenever
+/// appropriate, then after each of those calls, as soon as feasible call all
+/// polling methods (D) and deal with their outputs appropriately, e.g. by
+/// passing it to the application or by making a system-level I/O call. You
+/// should call the polling functions in this order:
+///
+/// 1. [`poll_transmit`](self::poll_transmit)
+/// 2. [`poll_timeout`](self::poll_timeout)
+/// 3. [`poll_endpoint_events`](self::poll_endpoint_events)
+/// 4. [`poll`](self::poll)
+///
+/// Currently the only actual dependency is from (2) to (1), however additional
+/// dependencies may be added in future, so the above order is recommended.
+///
+/// (A) may be called whenever desired.
+///
+/// Care should be made to ensure that the input events represent monotonically
+/// increasing time. Additionally we recommended that, for any given [`Instant`],
+/// you call [`handle_timeout`](self::handle_timeout) before you call
+/// [`handle_event`](self::handle_event) on any [`ConnectionEvent`]s having the same `Instant`.
```

Comment:
What's the basis for this recommendation?

Reply:
In case the logic for handling `Event(now = t)` ever implicitly depends on the timeout for `t` having already been fired, this ensures that this works consistently. (It would also be fine to say the timeout happens afterwards, as long as it's consistent.) It might not matter now and we could drop this, but I thought it'd be good to guard against future possibilities. It also gives something concrete to follow, in case the logic does need to be changed in the future, and the future developer doesn't know which way to go, but there's already different users assuming it differently.

Reply:
My inclination is to not specify requirements we don't have or expect to grow into, especially when it might be annoying for the caller to comply with them (e.g. requiring incoming events to be sorted). When there's no clear order between events, the logic should be robust to handling them in either order.

Reply:
OK, in that case I suggest stating as such in the documentation: i.e. "calling handle_event with events of same instant may be interleaved in any order with a call to handle_timeout at that same instant; however events or timeouts with different instants should not be interleaved", so callers know it's safe, and any future maintainers of quinn-proto don't change this behaviour?

What I've sometimes seen is people merging input blocking queues in a naive way, then this results in items being interleaved unpredictable and inappropriately at the consumer end; the latter "not be interleaved" requirement is supposed to expressly remind callers not to do that.

Reply:
I have no objection to a statement along those lines.